src/MOI_wrapper/objectives.jl

```diff
@@ -1,14 +1,40 @@
-# TODO: MOI.supports(::Optimizer, ::MOI.ObjectiveSense) = true
-MOI.supports(::Optimizer, ::MOI.ObjectiveFunction{ScalarFunction{F}}) where {F <: Function} = true
+MOI.supports(::Optimizer, ::MOI.ObjectiveSense) = true
+MOI.get(model::Optimizer, ::MOI.ObjectiveSense) = Min
+function MOI.set(model::Optimizer, ::MOI.ObjectiveSense, sense::MOI.OptimizationSense)
+    @warn "TODO: set sense" sense
+end
+
+"""
+    ScalarFunction(func, X)
+"""
+struct ScalarFunction{F <: Function, V <: Union{Nothing, VOV}} <: MOI.AbstractScalarFunction
+    f::F
+    X::V
+
+    ScalarFunction(f, X::Union{Nothing, VOV} = nothing) = (@warn X; new{typeof(f), typeof(X)}(f, X))
+end
+
+# external constructors
+
+function ScalarFunction(f, X::A) where {A <: AbstractArray{VariableRef}}
+    return ScalarFunction(f, VOV(vec(map(index, X))))
+end
+ScalarFunction(f, x::VariableRef) = ScalarFunction(f, [x])
+ScalarFunction(f, x::VI) = ScalarFunction(f, VOV([x]))
 
-# TODO: sense
-# MOI.get(model::Optimizer, ::MOI.ObjectiveSense) = model.inner.sense
+# copy
+Base.copy(func::ScalarFunction) = ScalarFunction(func.f, func.X)
 
-# function MOI.set(model::Optimizer, ::MOI.ObjectiveSense, sense::MOI.OptimizationSense)
-#     model.inner.sense = sense
-#     return
-# end
+# supports
+MOI.supports(::Optimizer, ::OF{ScalarFunction{F, V}}) where {F <: Function, V <: Union{Nothing, SVF,VOV}} = true
 
-function MOI.set(optimizer::Optimizer, ::MOI.ObjectiveFunction, func::ScalarFunction)
-    objective!(optimizer, func.f)
+# set
+function MOI.set(optimizer::Optimizer, ::OF, func::ScalarFunction{F, V}
+) where {F <: Function, V <: Union{Nothing, VOV}}
+    objective_function = if isnothing(func.X) # for direct use of MOI
```

Comment:
Best practice is to write two functions 
```Julia
function MOI.set(optimizer::Optimizer, ::OF{F}, func::F} where {F<:ScalarFunction{<:Function,Nothing}}
# and
function MOI.set(optimizer::Optimizer, ::OF{F}, func::F} where {F<:ScalarFunction{<:Function,VOV}}
```

Reply:
You're definitely right! This code is a mess ... I need to clean things up